core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -67,10 +70,13 @@ private ParallelIterator(
                             try (Closeable ignored =
                                 (iterable instanceof Closeable) ? (Closeable) iterable : () -> {}) {
                               for (T item : iterable) {
-                                queue.add(item);
+                                queue.put(item);
```

Comment:
@ConeyLiu I found `future.cancel` can't exit this loop, because `ConcurrentLinkedQueue` not check InterruptedException. So we need check `close` in this place, to avoid memory leak. 

Actural, I found trino kill query, but this loop will continue add to queue until finished.

We need use `LinkedBlockingQueue ` or add check iterator is `closed`, or both them.Just like below
```java
for (T item : iterable) {
      if (closed) {
         queue.clear()
         return;
     }
      queue.put(item);
}
```

Reply:
@Heltman can you please add a few more tests to `TestParallelIterable` that would exercise this exact condition you're describing?

Reply:
Close already clears the queue and cancels tasks. I don't think that we need to modify this.

Reply:
Cancels will not worker if use ConcurrentLinkedQueue. If we don't change to LinkedBlockingQueue, we need check closed every time before we add to queue.